### DISCUSSION TITLE
is https://on-air.nicegui.io/ not available?

### DISCUSSION THREAD

**Discussion by retsyo** [zauberzeug/nicegui#2089] [Q&A]
### Question

I read https://nicegui.io/documentation/section_configuration_deployment#nicegui_on_air, and find https://on-air.nicegui.io/ 

but there is nothing under this url

**Comment by rodja** [comment]
It should redirect you to https://on-air.nicegui.io/login where you can sign-up and later login. After successful auth you should be asked to create a organization and provide a device name.

**Comment by shubhamgandhi** [comment]
Still having a trouble using this feature.

I got the login, generated a token, and put it into my main file, example:
`ui.run(on_air="9DLNMOudoKHSz7Ra")
`

But where do I find the actually sharable URL? 

I tried different combinations of URL but all said "No such device".
https://on-air.nicegui.io/shubham/test/9DLNMOudoKHSz7Ra --> "test" is my default device name
https://on-air.nicegui.io/shubham/9DLNMOudoKHSz7Ra
https://on-air.nicegui.io/9DLNMOudoKHSz7Ra

I read through all the documentation too, but no luck. Can someone provide a little more detailed documentation for a coding newbie like me?

**Reply by rodja** [reply]
The URL should show up at launch on the console. In your case something like

```
NiceGUI ready to go on http://localhost:8080, and http://192.168.28.74:8080
NiceGUI is on air at http://on-air.nicegui.io/shubham/test/
```

**Reply by shubhamgandhi** [reply]
I thought that would have been the case but this is all I see, no on-air link:
`NiceGUI ready to go on http://localhost:8080, and http://192.168.8.221:8080
`

I'm using VS Code with the python3 interpreter and running the python environment within the application rather than on a terminal, in case this has anything to do with it.

Other thing I can think of is the device name: I think I just randomly picked "test" during the on-air signup process, because I wasn't sure which device name this should be -- one where I'm writing code, or trying to access the live web version. Anyhow, neither is named test, it's actually "Shubham's MacBook Pro 2". But I don't think a URL can have spaces like this, right?
Would this have something to do with it?

**Reply by rodja** [reply]
> Anyhow, neither is named test, it's actually "Shubham's MacBook Pro 2".

The device name is specified in the configuration on https://on-air.nicegui.com where you can also change it.

> I'm using VS Code with the python3 interpreter and running the python environment within the application rather than on a terminal, in case this has anything to do with it.

Strange. Could you try executing your code on a terminal? And does it also not appear if you use a very simple NiceGUI app (with only a label for example)?

**Reply by shubhamgandhi** [reply]
Still struggling with this :( 

I tried different combinations of device names in the configuration before pulling token, but no luck
Tried with computer name "Shubahm's MacBook Pro", tried it with hyphens since the localhost replaces spaces with hyphens, also tried with the name that shows up in the terminal prompt "Shubham's-MBP-2".

Tried with VSCode in-app terminal as well as regular terminal app, both result in the same message without "NiceGUI on air at .."
`shubhamgandhi@Shubhams-MBP-2 HobbyCoding % python3 loan_calculator.py
NiceGUI ready to go on http://localhost:8080, http://17.11.113.13:8080, and http://192.168.8.221:8080
`

When I log into the on-air website, I do a message at the top that says "Welcome Shubham Gandhi, your On Air account is free of charge until December 31, 2023!" -- could this have anything to do with? Access is disabled on backend based on the date?

Just to confirm the SOP -- I have the set the device name, then pull token, then update ui.run(on_air="<token>"), then run the .py file, correct?

Any advice on how "device name" text field is consumed, what on-air does with it?

**Reply by shubhamgandhi** [reply]
I also did try with a simple app, copy pasted one of the table examples from documentation. Added an on-air token and ran it but still nothing.

**Comment by rodja** [comment]
@shubhamgandhi Could you try again? In https://github.com/zauberzeug/nicegui/issues/2179 we fixed the deployment for Australia (and possibly for other regions, too).

**Reply by shubhamgandhi** [reply]
@rodja still no URL Tried with and without token. I updated to the latest NiceGUI package 1.4.9.
I'm in the US West region (San Jose, California)

**Comment by rodja** [comment]
I'm so sorry for the late reply @shubhamgandhi. I have not found the time to dig into this. Now I've tried it myself a bit more and noticed that indeed On Air is not working as expected when using a fresh installation.

```bash
python -c "from nicegui import ui; ui.label('test'); ui.run(on_air=True)"
```

Shows me the error `aiohttp package not installed` on each On Air reconnect attempt. If I install aiohttp via pip it starts working. Does that also fix it for you?

Btw, I tested the "fresh installation" with

```bash
docker run --rm -it python:3.11 /bin/sh -c 'pip install nicegui && python -c "from nicegui import ui; ui.label('\''test'\''); ui.run(on_air=True)"'
```

**Reply by rodja** [reply]
I prepared pull request #2426 for inclusion in next upcoming release. We already depended on `python-socketio` but apparently for the async client we need the package-extra `asyncio-client`.

**Reply by shubhamgandhi** [reply]
@rodja thanks for digging more.
I tried installing/upgrading all the packages you mentioned -- already had `aiohttp`, couldn't find `asyncio-client`, but installed `asyncio`, and upgraded `nicegui` to 1.4.12. 

Then tried your same command in terminal with a fresh install: `pip3 install nicegui && python3 -c "from nicegui import ui; ui.label('test'); ui.run(on_air=True)"`
It did not give me an error or show a URL :(

Also tried running it from a file `from nicegui import ui; ui.label('test'); ui.run(on_air=True)`
Same thing, no error or URL. Just shows the localhost IP location.

**Reply by shubhamgandhi** [reply]
If it helps here's my package list from `pip3 list`

```
Package            Version
------------------ ---------
aiofiles           23.2.1
aiohttp            3.9.1
aiosignal          1.3.1
annotated-types    0.5.0
anyio              3.7.1
async-timeout      4.0.3
asyncio            3.4.3
attrs              23.1.0
bidict             0.22.1
certifi            2023.7.22
charset-normalizer 3.2.0
click              8.1.7
contourpy          1.0.7
cycler             0.11.0
fastapi            0.109.0
fastapi-socketio   0.0.10
fonttools          4.39.3
frozenlist         1.4.0
future             0.18.3
h11                0.14.0
httpcore           0.17.3
httptools          0.6.0
httpx              0.24.1
idna               3.4
ifaddr             0.2.0
iso8601            1.1.0
itsdangerous       2.1.2
Jinja2             3.1.3
kiwisolver         1.4.4
markdown2          2.4.10
MarkupSafe         2.1.3
matplotlib         3.7.1
multidict          6.0.4
nicegui            1.4.12
nicegui-highcharts 1.0.1
numpy              1.24.2
orjson             3.9.5
packaging          23.0
pandas             1.5.3
Pillow             9.5.0
pip                23.3.2
pscript            0.7.7
pydantic           2.2.1
pydantic_core      2.6.1
Pygments           2.16.1
pyparsing          3.0.9
pyserial           3.5
python-dateutil    2.8.2
python-dotenv      1.0.0
python-engineio    4.8.1
python-multipart   0.0.6
python-socketio    5.11.0
pytz               2023.3
PyYAML             6.0
scipy              1.11.3
seaborn            0.12.2
serial             0.0.97
setuptools         65.5.0
simdkalman         1.0.3
simple-websocket   1.0.0
six                1.16.0
sniffio            1.3.0
starlette          0.35.1
termcolor          2.3.0
tsmoothie          1.0.4
typing_extensions  4.9.0
uvicorn            0.22.0
uvloop             0.17.0
vbuild             0.8.2
watchfiles         0.19.0
websockets         11.0.3
wsproto            1.2.0
yarl               1.9.2

```

**Reply by rodja** [reply]
What Python version are you running. And on what operating system?

**Reply by rodja** [reply]
I installed all the dependencies you listed in a Docker container, but could easily connect to the On Air service. Could you test in a clean Docker container with either:

```
docker run --rm -it python:3.11 /bin/sh -c 'pip install nicegui aiohttp && python -c "from nicegui import ui; ui.label('\''test'\''); ui.run(on_air=True)"'
```

Or Dockerfile

```dockerfile
FROM python:3.11

RUN pip install nicegui aiohttp

CMD python -c "from nicegui import ui; ui.label('test'); ui.run(on_air=True)"
```

and the commands 

```bash
docker build -t issue-2089 .
docker run --rm -it issue-2089
```

**Reply by shubhamgandhi** [reply]
I'm using python 3.11.5 on MacOS 13.1 Ventura.

This is my first time using docker, here's what I did: 

- downloaded the desktop app and opened it
- ran your script on terminal: 

    `docker run --rm -it python:3.11 /bin/sh -c 'pip install nicegui aiohttp && python -c "from nicegui import ui; ui.label('\''test'\''); ui.run(on_air=True)"'`

That seems to have generated a URL!
```
...
auto-reloading is only supported when running from a file
NiceGUI ready to go on http://localhost:8080, and http://172.17.0.2:8080
NiceGUI is on air at http://on-air.nicegui.io/devices/Hp5M7eJR/
```
But the URL doesn't seem to lead anywhere -- browser says Safari failed to open the page because the server unexpectedly dropped the connection.
The localhost and IP don't go anywhere either. It this expected since we're running in a container?

**Reply by rodja** [reply]
Yes, localhost and http://172.17.0.2:8080 are only valid inside the docker container (because we have not setup port mapping). But the On Air URL should show you the label "test". Could you try another browser and/or post the js console log output?

**Reply by shubhamgandhi** [reply]
Chrome worked!
<img width="397" alt="Screenshot 2024-01-26 at 3 41 17 PM" src="https://github.com/zauberzeug/nicegui/assets/6484391/a02e3874-5ee8-4a37-ad3f-b3f80f3e8c10">

Sorry where can I find the js console log output? Is it the console from the "Inspect" tab on the docker app?

[inspect_console_output.txt](https://github.com/zauberzeug/nicegui/files/14070169/inspect_console_output.txt)

**Reply by shubhamgandhi** [reply]
So should it always be run from docker? 
Or are we looking at some other issue here on why running python directly from terminal doesn't create an on-air URL?

**Reply by rodja** [reply]
Very interesting. I can reproduce the Safari issue and will look into it shortly. But still there is the issue that it's only working for you within docker. I'm running Python 3.11.7 on Mac OS 14.2.1 (Sonoma). But either should be fine and the error might be somewhere else. Let's dig deeper: Could you add a print statement (or breakpoint) in the NiceGUI code at this location to ensure it is called?

https://github.com/zauberzeug/nicegui/blob/8ad2757973ccb32db1f51d0c404ea8120ad3b72d/nicegui/air.py#L176

This is the loop which tries to establish the connection to the On Air server. It should normally establish the connection on the first try.

**Reply by shubhamgandhi** [reply]
I added a few print statements in this function, looks like it's getting hung up at the socket error.
Keeps re-attempting connection every 10 seconds or so but catches socket exception each time.

<img width="766" alt="Screenshot 2024-01-29 at 11 19 21 AM" src="https://github.com/zauberzeug/nicegui/assets/6484391/6d069683-b125-4da9-8e2f-968389094793">

**Reply by rodja** [reply]
We have found and fixed the issue you had with Safari, @shubhamgandhi. It was caused by a duplicated header entry we have send. Seems Safari (and curl) are treating it as an obscure error.

**Reply by rodja** [reply]
So lets dig deeper into the socket errors when connecting on your regular machine (not Docker). Here are a few things which would be interesting:
- remove the `websocket` transport
- use `log.exception('Could not connect to NiceGUI On Air server.')` instead of your `print("socket error?")` to see the exact error message and stacktrace
- print `RELAY_HOST` variable to ensure you are connecting to https://on-air.nicegui.io/

**Reply by shubhamgandhi** [reply]
I commented out the transport, added a print statement for relay host, and added the log exception:
```
        while True:
            try:
                print("awaiting connection")
                if self.relay.connected:
                    print("relay connected")
                    await self.relay.disconnect()
                await self.relay.connect(
                    f'{RELAY_HOST}?device_token={self.token}',
                    socketio_path='/on_air/socket.io',
                    # transports=['websocket', 'polling'],  # favor websocket over polling
                )
                break
            except socketio.exceptions.ConnectionError:
                print("Relay Host:")
                print(RELAY_HOST)
                print("Logging Exception:")
                log.exception('Could not connect to NiceGUI On Air server.') 
                pass

```


Here'e the console output:
```
NiceGUI ready to go on http://localhost:8080, http://17.11.54.188:8080, and http://172.16.53.20:8080
awaiting connection
Relay Host:
https://on-air.nicegui.io/
Logging Exception:
Could not connect to NiceGUI On Air server.
Traceback (most recent call last):
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/nicegui/air.py", line 178, in connect
    await self.relay.connect(
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/socketio/async_client.py", line 150, in connect
    raise exceptions.ConnectionError(exc.args[0]) from None
socketio.exceptions.ConnectionError: Cannot connect to host on-air.nicegui.io:443 ssl:True [SSLCertVerificationError: (1, '[SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1006)')]
```

Seems like an issue with verifying the SSL certificate. Hope it helps.

**Reply by rodja** [reply]
Well... our ssl certificate looks ok: https://www.sslshopper.com/ssl-checker.html#hostname=on-air.nicegui.io. ChatGPT lists these five possible causes:

1. Local Certificate Store: The user's system might have an outdated or corrupted certificate store. This can prevent the correct verification of SSL certificates. Updating the certificate store or manually adding your certificate to the trusted store could resolve this issue.
2. Python SSL Module: Since the user is on Python 3.11.7, it's possible that there's an issue with the SSL module in this particular version, especially if it works in a Docker container with a different Python version. Trying a different Python version on the user's machine could be a quick way to diagnose this.
3. Operating System Certificates: Mac OS might have specific SSL configurations or certificate stores that are not in sync with the Python environment. Ensuring that the OS certificates are up to date and properly configured might help.
4. Network Interference: Sometimes, local network configurations, such as firewalls or proxy servers, interfere with SSL connections. Checking for any such interference on the user's network could be insightful.
5. Python Environment: The issue might be within the user's Python environment. A corrupt installation of Python or certain libraries (like requests or socketio) could cause SSL issues. Reinstalling Python or the related libraries might help.

Could you maybe try a simple ssl request on the console or Python script which accesses https://on-air.nicegui.io/status? This should return a string "Ok".

**Reply by shubhamgandhi** [reply]
Sorry about delayed response 

I tried this which prints out "ok" as expected. Does that address cause #1 and #3?
```
import requests
response = requests.get('https://on-air.nicegui.io/status/')
print(response.text)

```
I tried on-air on multiple networks with the same result, so it's probably not network related (#4)
I will try to delete and re-install python to address cause #2 and #5.
Open to trying other things.

**Reply by OpenSourceIronman** [reply]
Is On Air working for anyone else in the Las Vegas, NV USA area? I get the following terminal error "On Air host "https://on-air.nicegui.io/" is not reachable. Please check your internet connection."

Running Python 3.13 and NiceGUI 2.24.2 in a venv with the following code. My internet is working fine as seem by Arc browser on the right side of the image below. 
```py
from nicegui import ui
import requests

response = requests.get('https://on-air.nicegui.io/status/')
print(response.text)

ui.label('NiceGUI On Air')
ui.run(native=True, on_air='VBGvymM2zVnRX0vr')
```

<img width="1920" height="1200" alt="image" src="https://github.com/user-attachments/assets/913d6095-badd-4258-90a9-d17a49e06d1e" />

**Reply by rodja** [reply]
We had no outage messages from that server. Have you tried from another machine/network to verify that it is not your connection?

**Comment by retsyo** [comment]
`on_air=True` does not create any on-air link at all

I'm using nicegui-1.4.18,  python-3.10.9.amd64 on my windows 10 64bits. I can opne https://on-air.nicegui.io/  with my browser without problem. However
```
python-3.10.9.amd64\python -c "from nicegui import ui; ui.label('test'); ui.run(on_air=True)" 
```
gives
```
auto-reloading is only supported when running from a file
NiceGUI ready to go on http://localhost:8080, http://10.33.17.137:8080, http://169.254.188.112:8080, http://169.254.64.143:8080, and http://169.254.92.194:8080
```
in other words, there is no on-air link

**Reply by rodja** [reply]
The message is just a not-so-well-written warning. I've just changed it to "disabling auto-reloading because is is only supported when running from a file", to make clear that everything is all right but only the auto-reload is not working due to not running from a file. I my test I've got a On Air url, so the question is why it does not work for you...
Which region are you located? Maybe I can find something in the server logs.

**Reply by retsyo** [reply]
I am in China mainland

**Reply by rodja** [reply]
Could you please try again. We have just fixed some internal redis connection errors.

**Comment by rodja** [comment]
@retsyo and everyone else: We have a major hick up in our service infrastructure. While it has already still solved in most regions a sustainable solution is still in the workings. We hope to make On Air stable in the next 48 hours.

**Reply by h0uter** [reply]
Hey is on air currently operational? When I launch on air and visit the url it loads my gui partially then I get the "No such device" error: 
<img width="680" alt="image" src="https://github.com/zauberzeug/nicegui/assets/40892865/214545d4-e277-4277-bc76-a46c70c77bbd">

the app works perfectly on localhost

**Reply by falkoschindler** [reply]
Thanks for letting us know, @h0uter! We'll look into it.

**Reply by rodja** [reply]
What region are you on? What is the output in the console when you start the app?

**Reply by h0uter** [reply]
europe, console shows no errors and the app works as expected when accessing it using localhost

**Reply by rodja** [reply]
Strange. Its working for our systems. Do you use an private token or are you using `ui.run(on_air=True)`? Do you see the message `NiceGUI is on air at https://europe.on-air.io/...` in the console?
